### PR TITLE
UINV-32 Invoice approval details set by system

### PR DESCRIPTION
**Pull request by AndreiShumski2**
## Purpose
`approvedDate` and `approvalBy` cannot be set manually. 
Should be set by the system
https://issues.folio.org/browse/UINV-32

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/invoices/InvoiceForm/InvoiceForm.js [hunk 1]
```diff
@@ -113,6 +113,10 @@ class InvoiceForm extends Component {
     const addressBillTo = get(find(addresses, { id: formValues.billTo }), 'address', '');
     const isEditPostApproval = IS_EDIT_POST_APPROVAL(initialValues.id, initialValues.status);
     const metadata = initialValues.metadata;
+    const users = get(parentResources, 'users.records', []);
+    const approvedByUser = get(users.find(user => user.id === get(initialValues, 'approvedBy')), 'personal', {});
+
+    initialValues.approvedByName = `${get(approvedByUser, 'firstName', '')} ${get(approvedByUser, 'lastName', '')}`;
```
Comment (aliaksei-chumakou) [inline src/invoices/InvoiceForm/InvoiceForm.js]: Please don't use initialValues for not-form inputs.

Comment (AndreiShumski2) [reply src/invoices/InvoiceForm/InvoiceForm.js]: Fixed

File: src/invoices/InvoiceDetails/InvoiceDetails.js [hunk 2]
```diff
@@ -32,6 +32,7 @@ class InvoiceDetails extends Component {
     onClose: PropTypes.func.isRequired,
     invoice: PropTypes.object.isRequired,
     deleteInvoice: PropTypes.func.isRequired,
+    parentResources: PropTypes.object.isRequired,
```
Comment (NikitaSedyx) [inline src/invoices/InvoiceDetails/InvoiceDetails.js]: please avoid usage of parent resources here

File: src/invoices/InvoiceDetails/InvoiceDetails.js [hunk 3]
```diff
@@ -78,9 +79,13 @@ class InvoiceDetails extends Component {
       deleteInvoice,
       onClose,
       invoice,
+      parentResources,
     } = this.props;
     const { sections, showConfirmDelete } = this.state;
     const vendorInvoiceNo = invoice.vendorInvoiceNo;
+    const users = get(parentResources, 'users.records', []);
```
Comment (NikitaSedyx) [inline src/invoices/InvoiceDetails/InvoiceDetails.js]: does it make sense to load only one user for related invoice?

Comment (AndreiShumski2) [reply src/invoices/InvoiceDetails/InvoiceDetails.js]: Yes. Fixed

File: src/invoices/InvoicesList/Invoice/InvoiceEditLayer.js [hunk 4]
```diff
@@ -35,6 +44,30 @@ class InvoiceEditLayer extends Component {
...
+    this.getUser();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.match.params.id !== prevProps.match.params.id) {
+      this.getUser();
+    }
+  }
+
+  getUser = () => {
```
Comment (NikitaSedyx) [inline src/invoices/InvoicesList/Invoice/InvoiceEditLayer.js]: function is copy paste

Comment (AndreiShumski2) [reply src/invoices/InvoicesList/Invoice/InvoiceEditLayer.js]: Created a common function

File: src/invoices/InvoicesList/Invoice/InvoiceDetailsLayer.js [hunk 5]
```diff
@@ -19,17 +21,44 @@ class InvoiceDetailsLayer extends Component {
...
     mutator: PropTypes.object.isRequired,
     onClose: PropTypes.func.isRequired,
     onEdit: PropTypes.func.isRequired,
+    parentResources: PropTypes.object.isRequired,
     resources: PropTypes.object.isRequired,
     showToast: PropTypes.func.isRequired,
   }
 
+  state = {
+    user: [],
```
Comment (NikitaSedyx) [inline src/invoices/InvoicesList/Invoice/InvoiceDetailsLayer.js]: array?

Comment (AndreiShumski2) [reply src/invoices/InvoicesList/Invoice/InvoiceDetailsLayer.js]: yes, an array of records (in this case with only one record)

File: src/invoices/InvoiceDetails/InvoiceDetails.js [hunk 6]
```diff
@@ -32,6 +32,11 @@ class InvoiceDetails extends Component {
     onClose: PropTypes.func.isRequired,
     invoice: PropTypes.object.isRequired,
     deleteInvoice: PropTypes.func.isRequired,
+    user: PropTypes.object,
```
Comment (NikitaSedyx) [inline src/invoices/InvoiceDetails/InvoiceDetails.js]: just user?

File: src/invoices/InvoicesList/Invoice/InvoiceEditLayer.js [hunk 7]
```diff
@@ -22,10 +25,26 @@ class InvoiceEditLayer extends Component {
       path: `${INVOICE_API}/:{id}`,
       throwErrors: false,
     },
+    user: {
+      ...USER,
+      GET: {
```
Comment (NikitaSedyx) [inline src/invoices/InvoicesList/Invoice/InvoiceEditLayer.js]: why not to have it in resources as `approvedByUser` resource or smth like that, copy paste

File: src/invoices/InvoiceDetails/InvoiceDetails.js (hunk 6 above)
Comment (AndreiShumski2) [reply src/invoices/InvoiceDetails/InvoiceDetails.js]: fixed

File: src/invoices/InvoicesList/Invoice/InvoiceEditLayer.js (hunk 7 above)
Comment (AndreiShumski2) [reply src/invoices/InvoicesList/Invoice/InvoiceEditLayer.js]: fixed

File: src/common/components/ApprovedByField/ApprovedByField.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+  return (
+    <KeyValue
+      label={<FormattedMessage id="ui-invoice.invoice.details.information.approvedBy" />}
+      value={approvedById ? approvedByName : ''}
+    />
+  );
+};
+
+ApprovedByField.manifest = Object.freeze({
```
Comment (aliaksei-chumakou) [inline src/common/components/ApprovedByField/ApprovedByField.js]: naming. I have doubt that it's a Field.

File: src/common/components/ApprovedByField/ApprovedByField.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  user: {
+    ...USERS,
+    GET: {
+      path: (queryParams, pathComponents, resourceValues, logger, props) => {
+        const approvedBy = props.approvedById;
+
+        if (approvedBy) {
+          return `${USERS_API}/${approvedBy}`;
```
Comment (aliaksei-chumakou) [inline src/common/components/ApprovedByField/ApprovedByField.js]: I think you can change the whole resource to this:
```
  approvedByUser: {
    ...BASE_RESOURCE,
    path: 'users/!{approvedByUserId}',
  },
```

File: src/common/components/ApprovedByField/ApprovedByField.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+        return null;
+      },
+    },
+  },
+});
+
+ApprovedByField.propTypes = {
+  resources: PropTypes.object.isRequired,
+  approvedById: PropTypes.object,
```
Comment (aliaksei-chumakou) [inline src/common/components/ApprovedByField/ApprovedByField.js]: Object? I doubt this.
I think it should more like 
```
ApprovedBy.propTypes = {
  // eslint-disable-next-line react/no-unused-prop-types
  approvedByUserId: PropTypes.string,
  resources: PropTypes.shape({
    approvedByUser: PropTypes.object,
  }).isRequired,
};
```

File: src/common/components/ApprovedByField/ApprovedByField.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import { stripesConnect } from '@folio/stripes/core';
+import { KeyValue } from '@folio/stripes/components';
+
+import { get } from 'lodash';
+
+import { USERS } from '../../resources';
+import { USERS_API } from '../../constants/api';
+
+const ApprovedByField = ({ approvedById, resources }) => {
+  const user = get(resources, ['user', 'records', 0]);
```
Comment (aliaksei-chumakou) [inline src/common/components/ApprovedByField/ApprovedByField.js]: I think you might miss the case when user is 404, and the previously loaded user will be displayed, what is the bug. Maybe it should look like 
```
const ApprovedBy = ({ resources }) => {
  const { failed, records } = resources.approvedByUser || {};
  const approvedByUser = failed ? null : get(records, '0');

  return (
    <KeyValue
      label={<FormattedMessage id="ui-invoice.invoice.approvedBy" />}
      value={getUserName(approvedByUser)}
    />
  );
};
```

File: src/common/components/ApprovedByField/ApprovedByField.js (hunk 9 above)
Comment (AndreiShumski2) [reply src/common/components/ApprovedByField/ApprovedByField.js]: Fixed

File: src/common/components/ApprovedByField/ApprovedByField.js (hunk 10 above)
Comment (AndreiShumski2) [reply src/common/components/ApprovedByField/ApprovedByField.js]: fixed

File: src/common/components/ApprovedByField/ApprovedByField.js (hunk 11 above)
Comment (AndreiShumski2) [reply src/common/components/ApprovedByField/ApprovedByField.js]: fixed. thanx

File: src/common/components/ApprovedByField/ApprovedByField.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  const approvedByName = `${get(user, 'personal.firstName', '')} ${get(user, 'personal.lastName', '')}`;
```
[the remark this answers is not in the document]
Comment (AndreiShumski2) [reply src/common/components/ApprovedByField/ApprovedByField.js]: fixed

File: src/common/components/ApprovedByField/ApprovedByField.js (hunk 8 above)
Comment (AndreiShumski2) [reply src/common/components/ApprovedByField/ApprovedByField.js]: Remaned

File: src/common/components/ApprovedBy/ApprovedBy.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+ApprovedBy.manifest = Object.freeze({
+  approvedByUser: {
+    ...BASE_RESOURCE,
+    path: `${USERS_API}/!{approvedByUserId}`,
+  },
+});
+
+ApprovedBy.propTypes = {
+  // eslint-disable-next-line react/no-unused-prop-types
```
Comment (aliaksei-chumakou) [inline src/common/components/ApprovedBy/ApprovedBy.js]: it's used, no need in this line

Comment (AndreiShumski2) [reply src/common/components/ApprovedBy/ApprovedBy.js]: yep)

**Comment by aliaksei-chumakou** [comment]
@AndreiShumski2 in case you've missed it - coverage was below 80% 

> 65.2% Coverage on New Code (is less than 80%)
It would be nice to check what lines are not covered and write tests. 
Also it's weird that checkbox regarding coverage is checked, and at the same time there is no explanation why it's merged without a proper coverage.

**Comment by AndreiShumski2** [comment]
This code was merged to unblock the possibility of creating invoices. The checkbox wasn't unchecked before merging my mistake. Added tests for covering new code. See [PR](https://github.com/folio-org/ui-invoice/pull/44).

**Pull request merged**